libherokubuildpack/src/buildpack_output/mod.rs

```diff
@@ -0,0 +1,819 @@
+    pub fn error(mut self, s: impl AsRef<str>) {
+        self.write_paragraph(&ANSI::Red, s);
+    }
+
+    /// Emit a warning message to the end user.
+    ///
+    /// A warning should be used to emit a message to the end user about a potential problem.
+    ///
+    /// Multiple warnings can be emitted in sequence. The buildpack author should take care not to overwhelm the end user with
+    /// unnecessary warnings.
+    ///
+    /// When emitting a warning, describe the problem to the user, if possible, and tell them how to fix it or where to look next.
+    ///
+    /// Warnings should often come with some disabling mechanism, if possible. If the user can turn off the warning,
+    /// that information should be included in the warning message. If you're confident that the user should not be able to
+    /// turn off a warning, consider using a [`BuildpackOutput::error`] instead.
+    ///
+    /// Warnings will be output in a multi-line paragraph style. A warning can be emitted from any state except for [`state::NotStarted`].
+    #[must_use]
+    pub fn warning(mut self, s: impl AsRef<str>) -> BuildpackOutput<S> {
+        self.write_paragraph(&ANSI::Yellow, s);
+        self
+    }
+
+    /// Emit an important message to the end user.
+    ///
+    /// When something significant happens but is not inherently negative, you can use an important message. For example,
+    /// if a buildpack detects that the operating system or architecture has changed since the last build, it might not be a problem,
+    ///, but if something goes wrong, the user should know about it.
+    ///
+    /// Important messages should be used sparingly and only for things the user should be aware of but not necessarily act on.
+    /// If the message is actionable, consider using a [`BuildpackOutput::warning`] instead.
+    #[must_use]
+    pub fn important(mut self, s: impl AsRef<str>) -> BuildpackOutput<S> {
+        self.write_paragraph(&ANSI::BoldCyan, s);
+        self
+    }
+
+    fn write_paragraph(&mut self, color: &ANSI, s: impl AsRef<str>) {
+        let io = self.state.write_mut();
+
+        if !io.was_paragraph {
+            writeln_now(io, "");
+        }
+
+        writeln_now(
+            io,
+            ansi_escape::wrap_ansi_escape_each_line(
+                color,
+                prefix_lines(s.as_ref(), |_, line| {
+                    // Avoid adding trailing whitespace to the line, if there was none already.
+                    // The `\n` case is required since `prefix_lines` uses `str::split_inclusive`,
+                    // which preserves any trailing newline characters if present.
+                    if line.is_empty() || line == "\n" {
+                        String::from("!")
+                    } else {
+                        String::from("! ")
+                    }
+                }),
+            ),
+        );
+        writeln_now(io, "");
```

Comment:
I think there might be a bug here - if there is a trailing newline in the input text to `write_paragraph`, then a final newline is still added here, causing doubled up newlines.

For example, if a modify one of the tests to do this:

```rust
            .warning("hello\nworld\n")
```

(note the trailing `\n\)

Then the output I get is like this:

```
            # Example Buildpack

            ! hello
            ! world


            - Done (finished in < 0.1s)
```

(with two newlines before the "done")

Now in the `.warning("hello\nworld\n")` example it seems like calling the function like that is user error, since they shouldn't add the trailing newline. In which case it would seem reasonable to not support this case.

However, I wonder whether in practice when people start using `build_output`, whether they will write things using `indoc!` like so:

```rust
            .start("Example Buildpack")
            .warning(indoc! {"
                Some intro/title

                A longer explanation of what is wrong...
            "})
            .finish();
```

Or like this:

```rust
        let message = indoc! {"
            Some intro/title

            A longer explanation of what is wrong...
        "};

        // ...
            .start("Example Buildpack")
            .warning(message)
            .finish();
```

And in both of those cases, the resultant message has a trailing newline, and the only way to prevent the trailing newline is to use `indoc!` like so:

```rust
            .start("Example Buildpack")
            .warning(indoc! {"
                Some intro/title

                A longer explanation of what is wrong..."
            })
            .finish();
```

(note the quote mark is now on the line before)

Or this:

```rust
        let message = indoc! {"
            Some intro/title

            A longer explanation of what is wrong..."
        };

        // ...
            .start("Example Buildpack")
            .warning(message)
            .finish();
```

I wonder if it might be easy for people to misplace the quote mark in this case, and not realise they are introducing an additional newline?

A fix might be to wrap this `writeln_now(io, "")` call in a `if !io.was_paragraph` check, like is used higher up in the `write_paragraph` implementation (for the paragraph's opening newline).

What are your thoughts? :-)

Reply:
Great catch. I believe it's the build output's responsibility to handle this. Similar to input sanitization of a form element in a web app. I fixed by trimming the input in shared functions for all inputs.

Reply:
Thank you for fixing this in 3ad2008e41dd1d4584142a8692f3aef677d2db96.

I spotted another case that wasn't tested (but already worked, so only needed updating the test):
https://github.com/heroku/libcnb.rs/pull/721/commits/aa19940d2b9c52eaad70468909eebf55c9b3e9ed

I also spotted what appears to be a limitation of the current `start_stream` / `line_mapped` / `MappedWrite` implementation. Specifically, if the streamed command's output has a trailing newline, then there are double newlines in the output. I think the fix for this will be more involved, so I would suggest we look at it later after merging this PR (but up to you). If you agree this is something we should fix longer term, would you mind filing a GitHub issue for it?

In the meantime I've added a testcase that demonstrates the issue (plus also tests the previously untested-but-working "newline in the middle of the output" case):
https://github.com/heroku/libcnb.rs/pull/721/commits/67e67aac889902924ad80d6a63b8b40fcd245940

Lastly, I spotted another small correctness/parity issue relating to streaming command output of blank lines (see commit message for more details) - I've fixed and updated the tests for it in:
https://github.com/heroku/libcnb.rs/pull/721/commits/d423789eee05fb9e108455ee64c0a403a7f25fdb

Reply:
>  I think the fix for this will be more involved, so I would suggest we look at it later after merging this PR (but up to you)

I pushed a fix that checks if a double newline was already added. We already had the infrastructure in place, it just wasn't being used. I modified your test case to remove that newline.

I'm good with everything else. Sorry for the delay this morning. I think we're good to go

Reply:
> I pushed a fix that checks if a double newline was already added. We already had the infrastructure in place, it just wasn't being used. I modified your test case to remove that newline.

Oh fantastic - glad it was an easy fix - the fix looks great :-)